### PR TITLE
Let's port to Synapse module API

### PR DESCRIPTION
**Pull request by Yoric**

### CODE REVIEW

**Review by turt2live — APPROVED** [review]
it looks like python, but I'm not the best person to review python. Would recommend seeking backend review as well.

File: src/commands/UnbanBanCommand.ts [hunk 1]
```diff
@@ -50,7 +50,7 @@ export async function parseArguments(roomId: string, event: any, mjolnir: Mjolni
     while (argumentIndex < 7 && argumentIndex < parts.length) {
         const arg = parts[argumentIndex++];
         if (!arg) break;
-        if (["user", "room", "server"].includes(arg.toLowerCase())) {
+        if (["user", "room", "server", "recentusers"].includes(arg.toLowerCase())) {
```
Comment (richvdh) [inline src/commands/UnbanBanCommand.ts:53]: I think it's probably best not to include unrelated changes in the same PR. If problems come up, it's harder to see what caused them, and to back out any problematic changes.

(Also, I don't understand what this is doing)

File: synapse_antispam/mjolnir/antispam.py [hunk 2]
```diff
@@ -42,7 +51,8 @@ def build_list(self, room_id):
 
     def get_list_for_room(self, room_id):
         if room_id not in self.rooms_to_lists:
-            self.rooms_to_lists[room_id] = BanList(api=self.api, room_id=room_id)
+            self.rooms_to_lists[room_id] = BanList(
+                api=self.api, room_id=room_id)
```
Comment (richvdh) [inline synapse_antispam/mjolnir/antispam.py:55]: recommend using black, for consistent formatting.

**Review by richvdh — CHANGES_REQUESTED** [review]
Remember you'll need to update the documentation for how the module should be installed.

It looks like this change will make the module incompatible with the old API. I recommend supporting both the new API at the same time as the old one, at least for a while, so that we don't need to deploy code and config changes at the same time.

One way to do that is to leave `AntiSpam` as it is today - complete with definitions of all the methods like `user_may_create_room` required to implement the old module api - and to define a new class (for example `mjolnir.Mjolnir`) - which implements the new api and just wraps `AntiSpam`.

File: src/commands/UnbanBanCommand.ts (hunk 1 above)
Comment (Yoric) [reply src/commands/UnbanBanCommand.ts:53]: No clue, looks like a merge error.

File: mx-tester.yml [hunk 3]
```diff
@@ -10,9 +11,22 @@ up:
...
+
 down:
   finally:
     - docker stop mjolnir-test-reverse-proxy || true
+
+modules:
+  - name: mjolnir
+    build:
+      - cp -r synapse_antispam $MX_TEST_MODULE_DIR
+    install:
```
Comment (Gnuxie) [inline mx-tester.yml:23]: This install step doesn't do anything and should be deleted. mx-tester already takes care of installing the module, this makes it hang.

**Review by Gnuxie — COMMENTED** [review]
Could we add the mx-tester build step to github actions?

**Comment by Gnuxie** [comment]
Ok, ofc we already have mx-tester build in the CI

File: synapse_antispam/mjolnir/antispam.py [hunk 4]
```diff
@@ -127,3 +136,24 @@ def user_may_publish_room(self, user_id, room_id):
...
+            user_may_invite=self.antispam.user_may_invite,
+            check_username_for_spam=self.antispam.check_username_for_spam,
+        )
+
+    # Spam-checker API
+    # Note that these are `async`, by opposition to the APIs in `AntiSpam`.
+    async def check_event_for_spam(self, event):
+        return self.antispam.check_event_for_spam(self, event)
+
+    async def user_may_invite(self, inviter_user_id, invitee_user_id, room_id):
```
Comment (richvdh) [inline synapse_antispam/mjolnir/antispam.py:155]: recommend including type annotations on new methods:

```suggestion
    async def user_may_invite(self, inviter_user_id: str, invitee_user_id: str, room_id: str) -> bool:
```

File: README.md [hunk 5]
```diff
@@ -99,25 +101,25 @@ pip install -e "git+https://github.com/matrix-org/mjolnir.git#egg=mjolnir&subdir
...
-    # display names by the entries in the user ban list. Default false.
-    block_usernames: false
-    # The room IDs of the ban lists to honour. Unlike other parts of Mjolnir,
-    # this list cannot be room aliases or permalinks. This server is expected
-    # to already be joined to the room - Mjolnir will not automatically join
-    # these rooms.
-    ban_lists:
-      - "!roomid:example.org"
+modules:
+  - module: mjolnir.AntiSpam
```
[the remark this answers is not in the document]
Comment (richvdh) [reply README.md:105]: Consider adding another import to https://github.com/matrix-org/mjolnir/blob/main/synapse_antispam/mjolnir/__init__.py, so that this can be referred to as `mjolnir.Module`.

File: synapse_antispam/mjolnir/antispam.py [hunk 6]
```diff
@@ -20,13 +20,16 @@
...
+#
+# Old API
+#
 class AntiSpam(object):
     def __init__(self, config, api):
         self.block_invites = config.get("block_invites", True)
         self.block_messages = config.get("block_messages", False)
         self.block_usernames = config.get("block_usernames", False)
         self.list_room_ids = config.get("ban_lists", [])
-        self.rooms_to_lists = {}  # type: Dict[str, BanList]
```
[the remark this answers is not in the document]
Comment (Yoric) [reply synapse_antispam/mjolnir/antispam.py:29]: Well, importing `Dict` isn't correct anymore, so I *guess* that this isn't either?
